[기본 레이아웃 컴포넌트] 첵스(최혜령) 미션 제출합니다.

HyeryongChoi opened the pull request:
# 🎯 1단계. 기본 레이아웃 컴포넌트

### [Npm 배포 링크](https://www.npmjs.com/package/chex-layout-component) | [Storybook 배포 링크](https://6505550eb343f2286f179ae2-huchuaarep.chromatic.com)

안녕하세요 도밥👋 방가방가~
이번 미션은 시간이 없어서 조금 힘들었지만 스토리북에 공을 들여봤습니ㄷr~ 편하게 리뷰 부탁드려요!

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## Container 컴포넌트
- [x] `minWidth`, `maxWidth` 속성 적용
- [x] npm 배포
- [x] README.md 작성

## Grid 컴포넌트
- [x] `rows`, `columns`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성

## Flex 컴포넌트
- [x] `direction`, `justify`, `align`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성

## 📌 구현한 내용 설명
<img src="https://github.com/woowacourse/layout-component/assets/24777828/df2103c0-b51e-4693-8f04-e632c291079c" width="400px" alt="" />

LMS에 나와있는 각 컴포넌트의 목적과 요구사항에 맞게 구현하려고 노력했습니다. 
그외 추가적인 속성은 필요할 것 같은 속성만 추가했어요. 구현하면서 고민했던 부분은 `minWidth`나 `maxWidth` 등의 값을 입력받을 때 `number`로 입력받을지 `string`으로 입력받을지 고민했습니다. emotion의 경우 숫자로 받으면 자동으로 `px`단위로 설정이 된다고 하던데 `styled-components`는 그렇진 않더라구요! 그래서 이번에는 단위 하나로 고정하기 보다는 `px, rem` 등 사용자가 원하는 단위로 쓸 수 있게 `string`으로 입력을 받도록 했습니다. 
그리고 npm 라이브러리 version을 어떤 기준으로 업데이트 해야할지 고민했습니다. 0.0.1 단위로 올리다가 완성됐다싶을 때 1버전으로 올렸는데요 이게 맞나~?싶네여 도밥의? 생각은? 어떠신지? 궁금합니다? 🤔

## 🔗 참고 링크 (선택 사항)
- [MUI Grid](https://mui.com/material-ui/react-grid/)
- [vite와 함께 리액트 컴포넌트 npm에 배포하기](https://devpluto.tistory.com/entry/React-vite%EC%99%80-%ED%95%A8%EA%BB%98-%EB%A6%AC%EC%95%A1%ED%8A%B8-%EC%BB%B4%ED%8F%AC%EB%84%8C%ED%8A%B8-npm%EC%97%90-%EB%B0%B0%ED%8F%AC%ED%95%98%EA%B8%B0)
- [가브리엘 블로그](https://leirbag.tistory.com/146)

wmakerjun committed:
docs: update guide for reviewer

HyeryongChoi committed:
chore: vite+react 프로젝트 세팅

HyeryongChoi committed:
chore: 배포자동화 설정

HyeryongChoi committed:
chore: 파일 경로 이동

HyeryongChoi committed:
chore: Storybook 설치

HyeryongChoi committed:
chore: chromatic 설치

HyeryongChoi committed:
chore: 스토리북(크로마틱) 자동배포 설정

HyeryongChoi committed:
fix: 크로마틱 배포 오류 해결 시도

HyeryongChoi committed:
chore: tsconfig 원복

HyeryongChoi committed:
chore: yarn 삭제

HyeryongChoi committed:
chore: styled-components 설치

HyeryongChoi committed:
feat: 글로벌스타일 적용

HyeryongChoi committed:
feat: 컨테이너 컴포넌트 추가

HyeryongChoi committed:
chore: 스토리북 배경색 설정

HyeryongChoi committed:
feat: Flex 컴포넌트 추가

HyeryongChoi committed:
docs: 스토리북 링크 추가

HyeryongChoi committed:
docs: 스토리북 링크 삭제

HyeryongChoi committed:
docs: JSDocs 추가

HyeryongChoi committed:
feat: 그리드 컴포넌트 추가

HyeryongChoi committed:
fix: 컨테이너 컴포넌트 스토리북 수정

HyeryongChoi committed:
chore: 불필요한 파일 삭제

HyeryongChoi committed:
chore: 파일 경로 이동

HyeryongChoi committed:
chore: 불필요한 파일 삭제

HyeryongChoi committed:
chore: 불필요한 파일 삭제

HyeryongChoi committed:
chore: 빌드 설정

HyeryongChoi committed:
chore: npm 배포 설정

HyeryongChoi committed:
fix: 오타 수정

HyeryongChoi committed:
chore: package.json 수정

HyeryongChoi committed:
chore: name 수정

HyeryongChoi committed:
feat: Container minHeight, maxHeight 속성 추가

HyeryongChoi committed:
docs: JSDocs 수정

HyeryongChoi committed:
feat: backgroundColor 속성 타입 변경

HyeryongChoi committed:
feat: 백그라운드컬러 기본값 변경

HyeryongChoi committed:
docs: JSDocs 수정

HyeryongChoi committed:
docs: README 업데이트

HyeryongChoi committed:
chore: 버전업데이트

HyeryongChoi committed:
chore: 1버전 업데이트

HyeryongChoi committed:
chore: 1.0.2버전

Creative-Lee commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,51 @@
+  /** 컴포넌트의 배경색상을 지정하는 속성입니다. */
+  backgroundColor?: CSSProperties['backgroundColor'];
+  /** 컨테이너에 적용할 CSS 스타일을 설정하는 속성입니다. */
+  css?: CSSProp;
```
현재 모든 컴포넌트의 css 속성이 의도대로 동작하지 않는 것 같아요!
![image](https://github.com/woowacourse/layout-component/assets/77152650/a2ee99d9-bdda-4fc1-b67c-afd8fb0ac74b)
위처럼 넣었는데,

![image](https://github.com/woowacourse/layout-component/assets/77152650/8baea1bb-33b9-4cc0-9ede-c4b870e0cb68)
 바인딩 되지 않고 있네요!

확인해 주세요 ㅎㅎ

HyeryongChoi replied:
```tsx
${({ css }) => (css ? `${css}` : '')}; // 수정 전

${({ css }) => css}; // 수정 후
```
아래처럼 수정했더니 적용이 되네요!! 도밥 덕분에 또 하나 더 배웠습니다~~~

Creative-Lee commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,51 @@
+import { CSSProp, CSSProperties, styled } from 'styled-components';
+
+interface Props extends ComponentPropsWithRef<ElementType> {
+  /** 컴포넌트의 HTML 태그를 지정하는 속성입니다. 예를 들어, tag="div"로 설정하면 컴포넌트가 \<div\> 태그로 렌더링됩니다. */
+  tag?: keyof JSX.IntrinsicElements;
```
💬 tag prop으로 사용하기로 한 태그의 attributes 도 자동완성 되면 좋을 것 같아요!
a 태그라면 href 같은 속성이 type 지원 되는거죠!

![tag props 어트리뷰트 미지원](https://github.com/woowacourse/layout-component/assets/77152650/830d217f-d0c6-4270-907f-a4c386220c70)

참고자료 드립니다~ ㅎㅎ 
[build-strongly-typed-polymorphic-components](https://blog.logrocket.com/build-strongly-typed-polymorphic-components-react-typescript/)

HyeryongChoi replied:
제네릭을 사용해서 아래와 같이 수정했을 때 속성 자동완성까지 나오는 것을 확인했습니다!
그런데 저의 경우 styled-components를 사용하는데 const Tag = styled.div<Props<any>>이 부분 때문에 빌드 시 에러가 나더라구요.. [참고](https://the-teacher.medium.com/typescript-type-instantiation-is-excessively-deep-and-possibly-infinite-how-to-fix-4b941cb28f4b) 이 글과도 관련이 있어보이는데 스컴에 특정한 prop 타입을 주지 못하고 `Props<any>`를 줘서 타입을 재귀적으로 탐색하게 되면서(?) 결국 터지는 것 같습니다. 정확한 원인은 잘 모르겠어요.. any말고 ElementType이나 keyof ReactHTML 등을 넣어봤는데 이 타입들을 넣었을 때는 타입 추론이 또 잘 안되는 문제가 있었습니다. 일단 그래서 이 부분은 다시 이전 코드로 돌려놓았어요. ㅠ,ㅠ 도밥이 말한 방법을 사용하려면 스컴 대신 기본 css나 이모션 등 다른 스타일링 방법을 사용해야할 것 같습니다!

<details>
  <summary>build 시 JavaScript heap out of memory 에러 나는 코드</summary>

  ```tsx
  import { ComponentPropsWithRef, ReactHTML } from 'react';
import { CSSProp, CSSProperties, styled } from 'styled-components';

type Props<T extends keyof ReactHTML> = ComponentPropsWithRef<T> & {
  /** 컴포넌트의 HTML 태그를 지정하는 속성입니다. 예를 들어, tag="div"로 설정하면 컴포넌트가 \<div\> 태그로 렌더링됩니다. */
  tag?: T;
  /** 컴포넌트의 최소 너비를 지정하는 속성입니다. 단위를 포함하여 입력해주세요.*/
  minWidth?: string;
  /** 컴포넌트의 최대 너비를 지정하는 속성입니다. 단위를 포함하여 입력해주세요.*/
  maxWidth?: string;
  /** 컴포넌트의 최소 높이를 지정하는 속성입니다. 단위를 포함하여 입력해주세요.*/
  minHeight?: string;
  /** 컴포넌트의 최대 높이를 지정하는 속성입니다. 단위를 포함하여 입력해주세요.*/
  maxHeight?: string;
  /** 컴포넌트의 배경색상을 지정하는 속성입니다. */
  backgroundColor?: CSSProperties['backgroundColor'];
  /** 컨테이너에 적용할 CSS 스타일을 설정하는 속성입니다. */
  css?: CSSProp;
};

const Container = <T extends keyof ReactHTML>(props: Props<T>) => {
  const { tag = 'div', ...restProps } = props;

  return (
    <Tag as={tag} {...restProps}>
      {restProps.children}
    </Tag>
  );
};

export default Container;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Tag = styled.div<Props<any>>`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  width: ${({ width }) => (width ? `${width}` : 'auto')};
  height: ${({ height }) => (height ? `${height}` : 'auto')};
  background-color: ${({ backgroundColor }) => (backgroundColor ? `${backgroundColor}` : 'transparent')};

  min-width: ${({ minWidth }) => (minWidth ? `${minWidth}` : 'auto')};
  max-width: ${({ maxWidth }) => (maxWidth ? `${maxWidth}` : 'none')};

  min-height: ${({ minHeight }) => (minHeight ? `${minHeight}` : 'auto')};
  max-height: ${({ maxHeight }) => (maxHeight ? `${maxHeight}` : 'none')};

  ${({ css }) => css};
`;
  ```

</details>

<details>
  <summary>에러코드</summary>

  ```tsx
  ❯ npm run build

> chex-layout-component@1.0.4 build
> tsc && vite build


<--- Last few GCs --->

[77515:0x140040000]    35284 ms: Mark-sweep 4036.1 (4141.0) -> 4025.0 (4140.0) MB, 2866.6 / 0.0 ms  (average mu = 0.480, current mu = 0.036) allocation failure; scavenge might not succeed
[77515:0x140040000]    38057 ms: Mark-sweep 4041.0 (4140.0) -> 4030.6 (4144.5) MB, 2760.5 / 0.0 ms  (average mu = 0.279, current mu = 0.004) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1048e0d00 node::Abort() [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 2: 0x1048e0ef0 node::ModifyCodeGenerationFromStrings(v8::Local<v8::Context>, v8::Local<v8::Value>, bool) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 3: 0x104a37854 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 4: 0x104be27d0 v8::internal::EmbedderStackStateScope::EmbedderStackStateScope(v8::internal::Heap*, v8::internal::EmbedderStackStateScope::Origin, cppgc::EmbedderStackState) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 5: 0x104be12b4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 6: 0x104bd555c v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 7: 0x104bd5d8c v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 8: 0x104bbbbc0 v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
 9: 0x104f4b740 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
10: 0x10529504c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
11: 0x1052222d0 Builtins_GrowFastSmiOrObjectElements [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
12: 0x10a74aea0 
13: 0x10a84d79c 
14: 0x10a8481a0 
15: 0x10a7487c8 
16: 0x10a3a6e60 
17: 0x10a7d4b64 
18: 0x10a69cab4 
19: 0x10a7fa61c 
20: 0x10a69c698 
21: 0x10a425cb4 
22: 0x10a8087b8 
23: 0x10a37b5e4 
24: 0x10a379d30 
25: 0x10a0f9394 
26: 0x10a6f4334 
27: 0x10a809850 
28: 0x10a37b5e4 
29: 0x10a379d30 
30: 0x10a6ec778 
31: 0x10a7e6e50 
32: 0x10a11aaa4 
33: 0x10a7e0d6c 
34: 0x10a3b939c 
35: 0x10a776f18 
36: 0x10a69d54c 
37: 0x10a7576c4 
38: 0x10a7e0934 
39: 0x10a3b939c 
40: 0x10a77703c 
41: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
42: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
43: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
44: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
45: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
46: 0x10a36b924 
47: 0x10a36af7c 
48: 0x10a369e64 
49: 0x10a36a55c 
50: 0x10a369fc0 
51: 0x10a36a25c 
52: 0x10a8772c0 
53: 0x10a876f9c 
54: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
55: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
56: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
57: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
58: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
59: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
60: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
61: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
62: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
63: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
64: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
65: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
66: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
67: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
68: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
69: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
70: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
71: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
72: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
73: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
74: 0x105220198 Builtins_InterpreterEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
75: 0x10521e4d0 Builtins_JSEntryTrampoline [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
76: 0x10521e164 Builtins_JSEntry [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
77: 0x104b63fa4 v8::internal::(anonymous namespace)::Invoke(v8::internal::Isolate*, v8::internal::(anonymous namespace)::InvokeParams const&) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
78: 0x104b634d8 v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
79: 0x104a5387c v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
80: 0x10494c510 node::Realm::ExecuteBootstrapper(char const*, std::__1::vector<v8::Local<v8::Value>, std::__1::allocator<v8::Local<v8::Value> > >*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
81: 0x1048ab37c node::StartExecution(node::Environment*, char const*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
82: 0x1048ab2d4 node::StartExecution(node::Environment*, std::__1::function<v8::MaybeLocal<v8::Value> (node::StartExecutionCallbackInfo const&)>) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
83: 0x104834584 node::LoadEnvironment(node::Environment*, std::__1::function<v8::MaybeLocal<v8::Value> (node::StartExecutionCallbackInfo const&)>) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
84: 0x10491cfd0 node::NodeMainInstance::Run() [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
85: 0x1048ae0f8 node::LoadSnapshotDataAndRun(node::SnapshotData const**, node::InitializationResult const*) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
86: 0x1048ae37c node::Start(int, char**) [/Users/hyeryong/.nvm/versions/node/v18.14.0/bin/node]
87: 0x109f19088 
sh: line 1: 77515 Abort trap: 6           tsc
  ```

</details>

Creative-Lee replied:
어우 너무 고생 많으셨습니다👍  
이런 시련을 드리려던건 아닌데 ㅠㅠ 미안해요
해당 코드와 에러 로그까지 다 첨부해주셔서 정말 감사합니다!
흐음... 참고 글도 읽어봤는데, 어렵네요🤣 우선 키워드를 알았으니, 다음에 시도해 봅시다...ㅎ
고생했어요!

Creative-Lee commented on src/lib/components/Container/Container.tsx:
```diff
@@ -0,0 +1,51 @@
+export default Container;
+
+const Tag = styled.div<Props>`
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+  align-items: center;
```
요소의 중앙 정렬을 위해 부여한 속성 맞나요~? ㅎㅎ
저는 사용자가 커스텀한 props외에 고정적으로 박아넣는 스타일을 최대한 줄여야 하나? 싶어서
`margin: 0 auto` 로 구현하였어요.   

💬 중앙 정렬을 위해 해당 속성을 선택한 이유가 있다면 가볍게 말해주세요 ㅎㅎ
단순히 궁금해서 코멘트 했습니다~

HyeryongChoi replied:
오 그냥 margin: 0 auto를 주는 방법도 있겠네요 저는 깊게 고민하지 않고 많이 사용하던 방식을 사용했습니다ㅎㅎ

Creative-Lee commented on src/lib/components/Container/Container.stories.tsx:
Container 컴포넌트의 배경색을 지정해봤는데, 표시가 안되는 현상이 !!
![image](https://github.com/woowacourse/layout-component/assets/77152650/1bb850e3-45ef-4e2c-97a9-b8250cc5e517)

![image](https://github.com/woowacourse/layout-component/assets/77152650/922860f0-e70e-4898-adaf-6a960689a675)

사용자가 스토리북을 보고 사용해볼 때 
첵스의 의도와 다르게 전달될것 같아서(헥스코드만 되더라고요) 코멘트 남깁니다 ㅎㅎ
***실제 컴포넌트에는 잘 되어요!**

HyeryongChoi replied:
ㅎㅎㅎ 사실 이건 스토리북에서 예쁘게 보여주려고 헥사코드를 받아서 rgba로 바꿔주도록 해서 요렇게 동작했던 것입니다...
```tsx
render: (args) => {
    return (
      <>
        <Container {...args} backgroundColor={hexToRgba(args.backgroundColor, 0.2) || '#bffff5'} />
        <Container {...args} backgroundColor={hexToRgba(args.backgroundColor, 0.4) || '#98fff0'} />
        <Container {...args} backgroundColor={hexToRgba(args.backgroundColor, 0.6) || '#64e3d0'} />
        <Container {...args} backgroundColor={hexToRgba(args.backgroundColor, 0.8) || '#49b9a8'} />
        <Container {...args} backgroundColor={args.backgroundColor || '#2c9484'} />
      </>
    );
  },
```

Creative-Lee commented on src/lib/components/Flex/Flex.stories.tsx:
```diff
@@ -0,0 +1,188 @@
+  render: (args) => {
+    return (
+      <>
+        <h4>display: {args.display}</h4>
```
이거 디테일 지리더라고요 ㄷㄷ 굿 👍😀

Creative-Lee commented on src/lib/components/Grid/Grid.tsx:
```diff
@@ -0,0 +1,64 @@
+   * 자동으로 생성되는 열의 크기를 지정하는 속성입니다. 문자열(ex. minmax(100px, auto))로 설정할 수 있습니다.
+   */
+  autoColumns?: string;
+  /**
+   * 그리드 아이템의 배치를 지정하는 속성입니다.
+   *  (ex. `"header header"
+   *  "sidebar content"
+   *  "footer footer"`)
+   */
+  areas?: string;
```
💬 해당 속성은 자식요소 (grid item)의 속성에도 필수로 area 속성이 적용되어야 하는데요~
제 경우에는 부모에는 props로 grid-template-areas 를 적용하고, 자식에는 css 속성을 별도로 넣어줘야하는게 어색하다고 생각했고. 어느정도 연계해서 제공해 주는게 좋겠다 싶어서 `Grid.Item` 컴포넌트를 만들어 두었어요.

책스는 어떤 의도로 해당 props(grid-template-areas속성)을 제공하기로 했나요~?

HyeryongChoi replied:
area 네이밍을 사용자가 선택해서 쓸 수 있도록 유연한 컴포넌트를 만들고싶었습니다! 

```tsx
<>
  <h4>
    Positioning with grid-template-areas
  </h4>
  <styled.div height="80rem">
    <Grid
      areas=`"a b b" "a c d" "a c d"`
      columns={3}
      gap="2rem"
    >
      <styled.div
        area="a"
        backgroundColor="#f6efff"
      >
        Item1
      </styled.div>
      <styled.div
        area="b"
        backgroundColor="#eee1ff"
      >
        Item2
      </styled.div>
      <styled.div
        area="c"
        backgroundColor="#e7d5ff"
      >
        Item3
      </styled.div>
      <styled.div
        area="d"
        backgroundColor="#def2ff"
      >
        Item4
      </styled.div>
    </Grid>
  </styled.div>
</>
```

Creative-Lee commented on src/lib/components/Flex/Flex.tsx:
```diff
@@ -0,0 +1,50 @@
+interface Props extends ComponentPropsWithRef<ElementType> {
+  /** 컴포넌트의 HTML 태그를 지정하는 속성입니다. 예를 들어, tag="div"로 설정하면 컴포넌트가 \<div\> 태그로 렌더링됩니다. */
+  tag?: keyof JSX.IntrinsicElements;
+  /** 컨테이너의 표시 방식을 지정하는 속성입니다. 'flex'는 블록요소, 'inline-flex'는 인라인요소로 설정합니다. */
+  display?: 'flex' | 'inline-flex';
+  /** 컨테이너의 주축(main axis)을 설정하는 속성입니다. 'row'는 수평 방향, 'column'은 수직 방향을 나타냅니다. 'row-reverse' 및 'column-reverse'는 주축 방향의 반대 방향을 나타냅니다. */
+  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
+  /**  주축(main axis)의 정렬 방법을 설정하는 속성입니다. 'flex-start'는 시작점 정렬, 'flex-end'는 끝점 정렬, 'center'는 가운데 정렬, 'space-between'은 주축에 여백을 고르게 분배하여 정렬, 'space-around'은 주축 양쪽에 여백을 고르게 분배하여 정렬, 'space-evenly'는 주축 양쪽에 동일한 여백을 고르게 분배하는 정렬입니다. */
+  justify?:
+    | 'flex-start'
+    | 'flex-end'
+    | 'center'
+    | 'space-between'
+    | 'space-around'
+    | 'space-evenly';
+  /** 교차축(cross axis)의 정렬 방법을 설정하는 속성입니다. 'flex-start'는 시작점 정렬, 'flex-end'는 끝점 정렬, 'center'는 가운데 정렬, 'stretch'는 아이템을 교차축에 맞게 늘리는 설정입니다. */
+  align?: 'flex-start' | 'flex-end' | 'center' | 'stretch';
+  /** Flex 아이템 사이의 간격을 지정하는 속성입니다. 단위를 포함하여 입력해주세요.*/
+  gap?: string;
```
💬직접 리터럴 유니온으로 타입을 지정하셨네요~
`CSSProperties['property']` 와 같은 기본 type 대신 적용하신 이유가 있으시다면 알려주세요~

HyeryongChoi replied:
몇가지 이유가 있는데요!
1. 각 속성의 모든 값들을 지원하지 않기 때문입니다.(Flex의 `display` 속성의 경우 `flex`, `inline-flex`만 지원)
2. 그리고 `CSSProperties['property']`를 사용하면 스토리북에서 사용자가 테스트할 때 string을 직접 입력해야해서 불편하겠더라구요!
  - <img width="787" alt="image" src="https://github.com/woowacourse/layout-component/assets/24777828/651ce111-f670-4bc3-9206-d3f5f52665ce">
    -> CSSProperties 타입으로 설정해도 아무 string이나 넣을 수 있어서? 그런 것 같습니다?
  - <img width="891" alt="image" src="https://github.com/woowacourse/layout-component/assets/24777828/61566de3-7237-4d5a-bc36-46aef6184ca6">
    -> 리터럴 유니온 타입으로 지정하면 이렇게 클릭 한번으로 선택이 가능합니다!
3. backgroundColor의 경우에는 `CSSProperties['backgroundColor']`을 사용했는데 그 이유는
백그라운드컬러의 경우 스토리북에서 컨트롤 설정을 바꿔줄 수 있기 때문입니다!
- ```tsx
  argTypes: {
      backgroundColor: { control: 'color' },
    },
  ```

Creative-Lee replied:
크으 타입 보셨군요!
[저도 세인과 해당 타입에 대해서 토론했는데](https://github.com/woowacourse/layout-component/pull/46#discussion_r1329603355), 한번 참고만 해보세요 ㅎㅎ
[해당 타입 비슷한 토론 2](https://github.com/woowacourse/layout-component/pull/46#discussion_r1330154788)

Creative-Lee commented on src/lib/components/Grid/Grid.stories.tsx:
```diff
@@ -0,0 +1,214 @@
+        <h4>Positioning with grid-auto-rows</h4>
+        <GridWrapper>
+          <Grid {...args}>
+            <Item backgroundColor='#f6efff'>Item1</Item>
+            <Item backgroundColor='#eee1ff'>Item2</Item>
+            <Item backgroundColor='#e7d5ff'>Item3</Item>
+            <Item backgroundColor='#def2ff'>Item4</Item>
+            <Item backgroundColor='#c3e7ff'>Item5</Item>
+            <Item backgroundColor='#b4e1ff'>Item6</Item>
```
공들인거 인정입니다 이쁘더라고요 ㅎ

Creative-Lee reviewed (changes requested):
# 📝 리뷰어를 위한 가이드라인

### 레이아웃 컴포넌트가 여러 환경에서도 잘 보이는지 확인해 주세요
- [x] 반응형: 레이아웃이 다양한 화면 크기에 적응하는지 확인해 주세요.
  - 내부적으로 breakpoint를 지정해 놓지 않았기에 통과입니다~ ㅎㅎ
- [x] 크로스 브라우징: 여러 브라우저(크롬, 사파리, 엣지 등)에서 동작하는지 확인해 주세요.
   - 속성들이 전부 기본적인 속성이라 모두 지원될거라고 생각합니다~
- [x] 첵스의 섬세한 cjs 지원
   - import, require까지 신경써주셨네요 ㄷㄷ 지립니다.
   - import는 사용되는거 확인했고, require는 모듈 불러오는거까지만 확인했습니다~ 환경 구성을 어떻게 해야할지 모르겠더라고요.. 허허 

### 레이아웃 컴포넌트를 사용하는 사용자로서, 잘 이해하고 활용할 수 있었는지 피드백해 주세요
- [x] 가독성: README나 사용법을 보고, 저자의 의도가 잘 이해되는지 독자의 입장에서 확인해 주세요.	
  - README에 테이블로 상세하게 적어주시고, props에 주석까지 잘 달려있어서 각 컴포넌트의 역할을 파악하기 쉬웠습니다.

- [x] 사용성: 레이아웃 컴포넌트를 사용하는 입장에서 쉽게 사용할 수 있었는지 확인해 주세요.
  - 처음에 number type으로 100 입력했는데 되지 않아서 당황했었는데,
	  **단위가 포함되어야 한다고 잘 알려주고 있어서** 스스로 꿀밤을 한 대 때렸습니다.

### 스타일 코드가 네이밍, 재사용성 등을 고려해 잘 작성되어 있는지 확인해 주세요
- [x] 의미론적 네이밍: 해당 CSS 클래스나 ID의 이름이 의미론적으로 적절한지 확인해 주세요.
	-  **스컴**의 특권으로 통과입니다ㅋㅋㅋ
- [x] 재사용성: 중복되거나 불필요한 스타일 규칙이 있는지 살펴보세요
	-  몇가지 궁금한 부분 코멘트 남겼습니다~

---

## 전하는 말
개인적으로 Layout컴포넌트를 설계하는 방식은 사람마다 전부 다를 거라고 생각해요.
오늘 수업 내용처럼 타겟 사용자가 다르기도 하고요.
각 컴포넌트를 설계하면서 속성, 기본값을 설정한 이유나 근거가 있다면 그만이지 않나~ 싶습니다
`때문에 이유나 생각을 묻는 코멘트가 많아졌네요~ 참고해주세요 ㅎㅎ`


## 첵스의 질문에 대한 대답😝

>`minWidth`나 `maxWidth`등의 값을 입력받을 때 `number`로 입력받을지 `string`으로 입력받을지 고민했습니다.

스타일드 컴포넌트는 px을 붙여주지 않나보군요?   
이와 별개로 첵스가 바라던게 number type도 지원하는 거라면, 충분히 구현해 볼 수 있을 것 같은데요~?ㅎㅎ 
number 입력시 기본 'px' postfix 기능을 지원해도 좋겠네요~ㅎㅎ 

> npm 라이브러리 version을 어떤 기준으로 업데이트 해야할지 고민했습니다.
> 0.0.1 단위로 올리다가 완성됐다싶을 때 1버전으로 올렸는데요 이게 맞나~? 싶네여 
> 도밥의? 생각은? 어떠신지? 궁금합니다? 🤔

첵스의? 질문에 대한? 저의? 생각은?
저 역시 첵스와 똑같은 흐름으로 했어요 ㅋㅋㅋㅋㅋㅋㅋ
조금 더 진지하게 생각해보면, 0.0.1 버전을 pubilsh 한 후부터 버전 관리가 시작되니깐,
사용자들에게 굳이 추한?!(동작하지 않는 상태 등등)모습을 보일 필요는 없다고 생각드네요..ㅎ

![image](https://github.com/woowacourse/layout-component/assets/77152650/71382873-54e6-4ab2-b5c7-88addf5a2d20)

HyeryongChoi committed:
fix:  스타일 적용 안되는 문제 해결

HyeryongChoi committed:
chore: 1.0.3버전 업데이트

HyeryongChoi committed:
feat: tag prop으로 사용하기로 한 태그의 attributes 타입 추론이 가능하도록 수정

- Flex경우 T extends ElementType을 사용했는데 타입추론이 되지 않음, 이유를 모르겠으나 비교를 위해 남겨놓음

HyeryongChoi committed:
chore: 1.0.4버전 업데이트

HyeryongChoi committed:
fix: 빌드 시 out of memory 에러로 원복

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory

styled-components와 타입 관련 문제인 듯

HyeryongChoi committed:
chore: 1.0.5버전

Creative-Lee commented:
괜히 설레발 쳐서 다형성 컴포넌트 적용해 보라고 해서 첵스 개고생 시켰네요...ㅎ 용서하세요
리뷰 반영 사항 다 확인했고, 답변 코멘트 2개 달아놨습니다. 
너무 고생많았고 화이팅이에용 스탭2에 봐용
![image](https://github.com/woowacourse/layout-component/assets/77152650/a545c7aa-d41e-472f-9191-c43b011c2d0c)
